core/src/main/java/org/lflang/generator/c/CReactionGenerator.java

```diff
@@ -343,11 +343,13 @@ public static String generateForwardBody(
                 + ")self->_lf__"
                 + actionName
                 + ".tmplt.token->value;",
+            "lf_critical_section_enter(self->base.environment);",
```

Comment:
I think the critical section needs to include the operations above, setting the value.  Also the setting of is_present below.